proxy/src/transport/tls/config.rs

```diff
@@ -367,3 +235,573 @@ fn set_common_settings(versions: &mut Vec<rustls::ProtocolVersion>) {
     // Only enable TLS 1.2 until TLS 1.3 is stable.
     *versions = vec![rustls::ProtocolVersion::TLSv1_2]
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ::fs_watch;
+    use task::test_util::BlockOnFor;
+
+    use tempdir::TempDir;
+    use tokio::runtime::current_thread::Runtime;
+
+    use std::{
+        path::Path,
+        io::Write,
+        fs::{self, File},
+    };
+    #[cfg(not(target_os = "windows"))]
+    use std::os::unix::fs::symlink;
+
+    use futures::{Sink, Stream};
+    use futures_watch::{Watch, WatchError};
+
+    struct Fixture {
+        cfg: CommonSettings,
+        dir: TempDir,
+        rt: Runtime,
+    }
+
+    const END_ENTITY_CERT: &'static str = "test-test.crt";
+    const PRIVATE_KEY: &'static str = "test-test.p8";
+    const TRUST_ANCHORS: &'static str = "ca.pem";
+
+    impl Fixture {
+        fn new() -> Fixture {
+            let _ = ::env_logger::try_init();
+            let dir = TempDir::new("certs").unwrap();
+            let cfg = CommonSettings {
+                trust_anchors: dir.path().join(TRUST_ANCHORS),
+                end_entity_cert: dir.path().join(END_ENTITY_CERT),
+                private_key: dir.path().join(PRIVATE_KEY),
+            };
```

Comment:
OK, now I can read the tests much better. I see that there's actually nothing TLS-specific about these tests except that they use `CommonSettings` to determine which files to look at, right?

Imagine that you replaced `cfg` with `paths: Vec<PathBuf>` or similar. Then all these tests would be completely independent of the `tls` submodule and they could all easily be moved to the fs_watch submodule. Not only that, but even more repetitive code in the tests could be replaced because you could just loop over the contents of `paths` (even now, they could be replaced with `for x in &[&fixture.cfg.trust_anchors, ...]` or similar).

Besides these tests that verify that the FS watching code works, in `tls/config.rs` (or, hopefully, somewhere under `tests/`) we're going to need to add tests that verify (1) When the input are initially not valid certificates/ca-bundles/private-keys, we ignore them, (2) When the inputs become valid, we construct a valid configuration from them, and (3) When we have a valid configuration and the files change to an invalid configuration, we ignore the change.

Reply:
> Besides these tests that verify that the FS watching code works, in tls/config.rs (or, hopefully, somewhere under tests/) we're going to need to add tests that verify (1) When the input are initially not valid certificates/ca-bundles/private-keys, we ignore them, (2) When the inputs become valid, we construct a valid configuration from them, and (3) When we have a valid configuration and the files change to an invalid configuration, we ignore the change.

^ Not in this PR, though.

Reply:
This is a fair point. I'll make the tests more generic and move them to the `fs_watch` module.